plugins/com.google.cloud.tools.eclipse.appengine.localserver.test/src/com/google/cloud/tools/eclipse/appengine/localserver/ui/ServerPortExtensionTest.java

```diff
@@ -0,0 +1,284 @@
+    assertEquals("8080", portExtension.portText.getText());
+  }
+
+  @Test
+  public void testPortText_numericCharacters() {
+    portExtension.portText.setText("808");
+    assertEquals("808", portExtension.portText.getText());
+
+    portExtension.portText.setText("8108");
+    assertEquals("8108", portExtension.portText.getText());
+  }
+
+  @Test
+  public void testPortText_allowArbitrarilyLargeNumbers() {
+    portExtension.portText.setText("80808");
+    assertEquals("80808", portExtension.portText.getText());
+
+    portExtension.portText.setText("480808");
+    assertEquals("480808", portExtension.portText.getText());
+
+    portExtension.portText.setText("9480808");
+    assertEquals("9480808", portExtension.portText.getText());
+  }
+
+  @Test
+  public void testAttributeSetOnValueChange() {
+    portExtension.portText.setText("12345");
+    assertEquals("12345", portExtension.portText.getText());
+    verify(server).setAttribute("appEngineDevServerPort", 12345);
+
+    portExtension.portText.setText("54321");
+    assertEquals("54321", portExtension.portText.getText());
+    verify(server).setAttribute("appEngineDevServerPort", 54321);
+
+    portExtension.portText.setText("0");
+    assertEquals("0", portExtension.portText.getText());
+    verify(server).setAttribute("appEngineDevServerPort", 0);
+
+    portExtension.portText.setText("987654321");
+    assertEquals("987654321", portExtension.portText.getText());
+    verify(server).setAttribute("appEngineDevServerPort", 987654321);
+  }
+
+  @Test
+  public void testAttributeDoesNotChangeOnBlockingNonNumericCharacters() {
+    portExtension.portText.setText("12345");
+    portExtension.portText.setText("12A345");
+    assertEquals("12345", portExtension.portText.getText());
+    verify(server, times(1)).setAttribute("appEngineDevServerPort", 12345);
+  }
+
+  @Test
+  public void testAttributeSetToZeroOnEmptyField() {
+    portExtension.portText.setText("");
+    assertTrue(portExtension.portText.getText().isEmpty());
+    verify(server).setAttribute("appEngineDevServerPort", 0);
+  }
+
+  @Test
+  public void testNoPopUpDecorationForValidPorts() {
+    portExtension.portText.setVisible(true);
+
+    portExtension.portText.setText("0");
+    assertFalse(portExtension.portDecoration.isVisible());
+
+    portExtension.portText.setText("1024");
+    assertFalse(portExtension.portDecoration.isVisible());
+
+    portExtension.portText.setText("8080");
+    assertFalse(portExtension.portDecoration.isVisible());
+
+    portExtension.portText.setText("65535");
+    assertFalse(portExtension.portDecoration.isVisible());
+  }
+
+  @Test
+  public void testShowErrorDecorationOnInvalidPorts() {
+    portExtension.portText.setVisible(true);
+
+    portExtension.portText.setText("65536");
+    assertTrue(portExtension.portDecoration.isVisible());
+    assertEquals(portExtension.errorImage, portExtension.portDecoration.getImage());
+
+    portExtension.portText.setText("65535");
```

Comment:
If the point here to verify that the decorator disappears if the invalid value is changed to a valid, then I'd move this case into a separate test method.